packages/structure/src/x/vscode.ts

```diff
@@ -35,3 +42,276 @@ class VSCodeWindowMethodsWrapper implements VSCodeWindowMethods {
+type TreeItemMenu = MenuCLI | MenuRoute | MenuGroup | MenuWithDoc
+
+interface MenuCLI {
+  kind: 'cli'
+  doc?: Command
+  run: Command
+}
+
+interface MenuRoute {
+  kind: 'route'
+  openInBrowser?: Command
+  openComponent?: Command
+  openRoute?: Command
+}
+
+interface MenuGroup {
+  kind: 'group'
+  add?: Command
+  doc?: Command
+}
+
+interface MenuWithDoc {
+  kind: 'withDoc'
+  doc?: Command
+}
+
+/**
+ * Based on the actual TreeItem interface provided by VSCode.
+ * It has a few differences.
+ */
+export type TreeItem2 = Partial<SerializableTreeItem> & {
+  key?: string
+  children?(): vscode.ProviderResult<TreeItem2[]>
+}
+
+export class TreeItem2Wrapper {
+  constructor(
+    public item: TreeItem2,
+    public parent?: TreeItem2Wrapper,
+    public indexInParent: number = 0
+  ) {}
+  @lazy() get keys(): string[] {
+    if (!this.parent) return []
+    return [...(this.parent?.keys ?? []), this.key]
+  }
+  @lazy() get key(): string {
+    const {
+      indexInParent,
+      item: { key, label },
+    } = this
+    if (key) return key
+    return (label ?? '') + '-' + indexInParent
+  }
+  @lazy() get id() {
+    return JSON.stringify(this.keys)
+  }
+  @lazy() get collapsibleState() {
+    return (
+      this.item.collapsibleState ??
+      (this.item.children
+        ? TreeItemCollapsibleState2.Collapsed
+        : TreeItemCollapsibleState2.None)
+    )
+  }
+  @memo() async children(): Promise<TreeItem2Wrapper[]> {
+    const cs = await ProviderResult_normalize(this.item.children?.())
+    return (cs ?? []).map((c, i) => new TreeItem2Wrapper(c, this, i))
+  }
+  @memo()
+  async findChild(key: string): Promise<TreeItem2Wrapper | undefined> {
+    for (const c of await this.children()) if (c.key === key) return c
+  }
+  @memo(JSON.stringify)
+  async findChildRec(keys: string[]): Promise<TreeItem2Wrapper | undefined> {
+    if (keys.length === 0) return this
+    const [k, ...rest] = keys
+    return await (await this.findChild(k))?.findChildRec(rest)
+  }
+  @lazy() get serializableTreeItem(): SerializableTreeItem {
+    return {
+      ...this.item,
+      id: this.id,
+      collapsibleState: this.collapsibleState,
+    }
+  }
+}
+
+/**
+ * https://microsoft.github.io/vscode-codicons/dist/codicon.html
+ * plust a few extra icons provided by decoupled studio:
+ * - redwood
+ * - prisma
+ * - graphql
+ * - netlify
+ */
+type ThemeIcon2 = string
+
+/**
+ * A copy of vscode.TreeItemCollapsibleState
+ * we don't want to have a runtime dependency on the vscode package
+ */
+export enum TreeItemCollapsibleState2 {
+  /**
+   * Determines an item can be neither collapsed nor expanded. Implies it has no children.
+   */
+  None = 0,
+  /**
+   * Determines an item is collapsed
+   */
+  Collapsed = 1,
+  /**
+   * Determines an item is expanded
+   */
+  Expanded = 2,
+}
+
+/**
+ * A vscode.TreeDataProvider that uses string IDs as elements
+ * and returns a SerializableTreeItem.
+ */
+type RemoteTreeDataProvider = ReplacePropTypes<
```

Comment:
The outline is now almost completely based on the vscode.TreeDataProvider API. I only changed it slightly to allow it to be called over the wise. I eliminated any abstractions previously introduced by Decoupled Studio. This results in better performance but, most importantly, it makes things more transparent.